### PR TITLE
Add: Public Wallets

### LINKED ISSUE #470: When user shared the wallet, show the other user public wallet
**Expectation**

I click on the Share icon on my profile and share link with user B.
User B clicks the link and lands on my Wallet view where he/she can see all the tokens I hold and all the token balances I have.
This means, we need to create a param link with all the public wallet information of the user.

### PR DESCRIPTION
**Pull request by joel-medicala-yral**
closes #470

### CODE REVIEW

File: ssr/src/page/token/info.rs [hunk 1]
```diff
@@ -131,7 +151,12 @@ pub fn TokenInfo() -> impl IntoView {
 
         let cans = unauth_canisters();
         let meta = token_metadata_by_root(&cans, params.user_principal, params.token_root).await?;
-        Ok(meta.map(|m| (m, (params.token_root, params.user_principal))))
+        let root = cans.sns_root(params.token_root).await;
+        let sns_cans = root.list_sns_canisters(ListSnsCanistersArg {}).await?;
+        let Some(ledger) = sns_cans.ledger else {
+            return Err(ServerFnError::new("Couldn't find ledger"));
+        };
+        Ok(meta.map(|m| (m, (params.token_root, params.user_principal), ledger)))
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/token/info.rs:159]: You can add a new field for the ledger principal here: https://github.com/yral-dapp/hot-or-not-web-leptos-ssr/blob/0bddfeb068acedb52fbe3d5b3502ea417b7e78b8/ssr/src/utils/token/mod.rs#L249

File: ssr/src/page/wallet/mod.rs [hunk 2]
```diff
@@ -36,7 +38,14 @@ fn ProfileGreeter(details: ProfileDetails) -> impl IntoView {
 
     view! {
         <div class="flex flex-col">
-            <span class="text-white/50 text-md">Welcome!</span>
+            {move || {
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/wallet/mod.rs:41]: This does not need to be a closure
```
{is_own_account.then(|| {
     ...
})}
```
Should work fine

File: ssr/src/component/infinite_scroller.rs [hunk 3]
```diff
@@ -15,6 +15,16 @@ pub trait KeyedData {
...
+pub(crate) trait KeyedCursoredDataProvider<T>:
+    crate::component::infinite_scroller::CursoredDataProvider
+{
+    async fn get_by_cursor_by_key(
+        &self,
+        start: usize,
+        end: usize,
+        user: T,
+    ) -> Result<PageEntry<Self::Data>, Self::Error>;
+}
```
Comment (rupansh-sekar-yral) [inline ssr/src/component/infinite_scroller.rs:27]: As opposed to an entirely new trait, `TokenRootList` can become
```
struct TokenRootList {
    canisters: Canisters<false>,
    user_canister: Principal
}
```

Then you can just use `canisters.get_individual_canister_by_user_principal(self.user_canister)` inside the `CursoredDataProvider` implementation
You can update the history provider implementation similarly

File: ssr/src/page/wallet/mod.rs [hunk 4]
```diff
@@ -116,43 +131,122 @@ fn TokensFetch() -> impl IntoView {
...
+                    }
+                }
+            }
+        }
+    }
+}
+#[component]
+pub fn WalletImpl(principal: Principal) -> impl IntoView {
     let (is_connected, _) = account_connected_reader();
+    let param_principal = move || principal;
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/wallet/mod.rs:168]: why do you need this?

File: ssr/src/page/wallet/mod.rs [hunk 5]
```diff
@@ -116,43 +131,122 @@ fn TokensFetch() -> impl IntoView {
...
 
     let auth_cans = authenticated_canisters();
-    let balance_fetch = auth_cans.derive(
-        || (),
-        |cans_wire, _| async move {
-            let cans = cans_wire?.canisters()?;
-            let user = cans.authenticated_user().await;
+    let balance_fetch = auth_cans.derive(param_principal, |cans_wire, principal| async move {
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/wallet/mod.rs:171]: Do you need an authenticated canister for this?
If not, use unauth_canisters with create_resource

File: ssr/src/utils/profile.rs [hunk 6]
```diff
@@ -57,7 +57,9 @@ impl ProfileDetails {
             .clone()
             .unwrap_or_else(|| self.principal.to_text())
     }
-
+    pub fn principal(&self) -> String {
```
Comment (rupansh-sekar-yral) [inline ssr/src/utils/profile.rs:60]: nit: spacing between function declerations

**Review by rupansh-sekar-yral — CHANGES_REQUESTED** [review]
Also run `leptosfmt ssr/src` after you're done
https://github.com/bram209/leptosfmt

File: ssr/src/page/wallet/mod.rs (hunk 5 above)
Comment (joel-medicala-yral) [reply ssr/src/page/wallet/mod.rs:171]: do i do the same for history fetch too?

**Review by rupansh-sekar-yral — APPROVED** [review]

**Pull request merged**